### PR TITLE
Added parameter entry points

### PR DESCRIPTION
**Pull request by santacodes**
Additions - 
+ `parameter_sets.py` from the main PyBaMM repository which acts as an entry point for the parameter sets mapped through the group `cookie_parameter_sets`.
+ `Chen2020.py` to test and load a parameter set through the entry point to ensure that everything works. 
+ Tests to test the entry points.

Updates - 
- A nox `dev` session. 
- Updated rules in compliance with the main PyBaMM repository in `pyproject.toml`.

Accessing parameter sets through entry points here is the same as accessing it through PyBaMM. Instead using `pybamm.parameter_sets` here it would imply `pybamm_cookiecutter.paramter_sets` to access them. I think it would make it easier for current users to adapt to the template as the implementations are almost the same. Before release, the template should be modified in compliance with the module name a user would give.

### CODE REVIEW

**Comment by santacodes** [comment]
@arjxn-py @brosaplanella could you please review this PR?

File: src/pybamm_cookiecutter/parameters/parameter_sets.py
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:1]: Since the code for this file is derived that in PyBaMM, we shall need to PyBaMM's BSD-3-Clause license and provide attribution. Could you add a comment about this, add the license based on the template at https://opensource.org/license/bsd-3-clause, and add a `LICENSES-bundled.txt` file in the project root with: the path to the file, the SPDX license identifier (BSD-3-Clause in this case), and the name of the project (PyBaMM)?

This license must be included in our distribution for this project in source code (sdist) and binary form (wheels) respectively. You might need to browse the documentation in https://hatch.pypa.io/latest/plugins/builder/sdist/#default-file-selection and build these distribution formats (`pip install build` and `python -m build --sdist --wheel` to ensure that `LICENSES-bundled` is present.

Comment (agriyakhetarpal) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:1]: and similarly for `Chen2020.py` (though I am not sure right now about what the associated paper at https://iopscience.iop.org/article/10.1149/1945-7111/ab9050/pdf/ about the experimental data)

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            # this file now return the name as a string, which will load the same
+            # parameter set as before when passed to `ParameterValues`
+            if name in self:
+                msg = (
+                    f"Parameter sets should be called directly by their name ({name}), "
+                    f"instead of via pybamm_cookiecutter.parameter_sets (pybamm_cookiecutter.parameter_sets.{name})."
+                )
+                warnings.warn(msg, DeprecationWarning, stacklevel=2)
+                return name
+            raise error
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:98]: Since we will be supporting recent versions of PyBaMM – say, `>=23.5`, perhaps we can remove this check for `pybamm-cookiecutter`'s case here? @brosaplanella

File: tests/test_entry_points.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import glob
+import importlib.util
+import sys
+
+def test_entry_points():
+    """Testing if the entry points are loading right"""
+
+    entry_points = list(pybamm_cookiecutter.parameter_sets)
+    parameter_sets = glob.glob("src/pybamm_cookiecutter/parameters/input/*.py")
+    # Making a list Parameter sets in the parameters/input directory
```
Comment (agriyakhetarpal) [inline tests/test_entry_points.py:12]: Might be easier to pass a `pathlib.Path` here or use `Path.glob()`

File: pyproject.toml [hunk 3]
```diff
@@ -92,32 +98,53 @@ port.exclude_lines = [
...
+    # "PL",          # pylint
+    # "PT",          # flake8-pytest-style
+    # "PTH",         # flake8-use-pathlib
+    # "RET",         # flake8-return
+    "RUF",           # Ruff-specific
+    # "SIM",         # flake8-simplify
+    # "T20",         # flake8-print
+    "UP",            # pyupgrade
+    "YTT",           # flake8-2020
+    "TID252",        # relative-imports
```
Comment (brosaplanella) [inline pyproject.toml:128]: Do we need to keep the comment ones or can we delete them?

File: src/pybamm_cookiecutter/parameters/parameter_sets.py
Comment (brosaplanella) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:1]: You mean to give attribution for the parameter set?

File: src/pybamm_cookiecutter/parameters/parameter_sets.py (hunk 1 above)
Comment (brosaplanella) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:98]: Given that the cookiecutter is under development, I think it is fine to only support the most recent version(s) for now.

**Review by brosaplanella — COMMENTED** [review]
Nice work, added some additional comments to Agriya's.

File: src/pybamm_cookiecutter/parameters/parameter_sets.py
Comment (agriyakhetarpal) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:1]: Yes. The paper says that the work is distributed under CC BY-NC-ND, but the code that PyBaMM uses to provide the work as a parameter set is under BSD, hence the non-clarity

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    def __new__(cls):
+        """Ensure only one instance of ParameterSets exists"""
+        if not hasattr(cls, "instance"):
+            cls.instance = super().__new__(cls)
+        return cls.instance
+
+    def __getitem__(self, key) -> dict:
+        return self.__load_entry_point__(key)()
+
+    def __load_entry_point__(self, key) -> Callable:
```
Comment (arjxn-py) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:90]:
```suggestion
    def _load_entry_point(self, key) -> Callable:
```
double underscores are not typical for non-magic methods. I'd suggest it renaming as a private method instead.

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    def get_docstring(self, key):
+        """Return the docstring for the ``key`` parameter set"""
+        return textwrap.dedent(self.__load_entry_point__(key).__doc__)
+
+    def __getattribute__(self, name):
+        try:
+            return super().__getattribute__(name)
+        except AttributeError as error:
+              raise error
```
Comment (arjxn-py) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:117]:
```suggestion
    def __getattribute__(self, name: str) -> Any:
        try:
            return super().__getattribute__(name)
        except AttributeError as error:
            if name in self:
                msg = (
                    f"Parameter sets can be called directly by their name ({name}), "
                    f"instead of via pybamm_cookiecutter.parameter_sets (pybamm_cookiecutter.parameter_sets.{name})."
                )
                warnings.warn(msg, DeprecationWarning, stacklevel=2)
                return name
            raise error
```
Does this look better? Make sure to test it with edge cases too.

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        Parameters for the Enertech cell (Ai2020), from the papers :footcite:t:`Ai2019`,
+        :footcite:t:`rieger2016new` and references therein.
+        ...
+
+    See also: :ref:`adding-parameter-sets`
+
+    """
+
+    def __init__(self):
+        # Dict of entry points for parameter sets, lazily load entry points as
```
Comment (arjxn-py) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:68]:
```suggestion
        '''Dict of entry points for parameter sets, lazily load entry points as'''
```
Is this docstring complete?

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        self.__all_parameter_sets = dict()
+        for entry_point in self.get_entries("cookie_parameter_sets"):
+            self.__all_parameter_sets[entry_point.name] = entry_point
+
+    @staticmethod
+    def get_entries(group_name):
+        # Wrapper for the importlib version logic
```
Comment (arjxn-py) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:75]:
```suggestion
        '''Wrapper for the importlib version logic'''
```

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        if sys.version_info < (3, 10):  # pragma: no cover
```
Comment (arjxn-py) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:76]: You can add testing for this line by maybe manipulating `sys.version_info` in tests?

File: tests/test_entry_points.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import pytest
+import pybamm_cookiecutter
+import importlib.util
+import sys
+from pathlib import Path
+
+def test_entry_points():
+    """Testing if the entry points are loading right"""
```
Comment (arjxn-py) [inline tests/test_entry_points.py:8]:
```suggestion
    """Test if the entry points are loaded correctly."""
```
Nitpicking 😅

File: noxfile.py [hunk 10]
```diff
@@ -1,9 +1,13 @@
...
+from pathlib import Path
+import os
 
 # Options to modify nox behaviour
 nox.options.default_venv_backend = "uv|virtualenv"
 nox.options.reuse_existing_virtualenvs = True
 
+VENV_DIR = Path("./venv").resolve()
+
 @nox.session(name="docs")
```
Comment (arjxn-py) [inline noxfile.py:11]:
```suggestion
@nox.session(name="docs", reuse_venv=True)
```

File: noxfile.py [hunk 11]
```diff
@@ -26,3 +30,11 @@ def run_template_generation(session):
     session.install("setuptools", silent=False)
     session.install("-e", ".[dev]", silent=False)
     session.run("pytest", "tests")
+
+@nox.session(name="dev")
```
Comment (arjxn-py) [inline noxfile.py:34]:
```suggestion
@nox.session(name="dev", reuse_venv=True)
```
I think it should be good to have `reuse_venv=True` in sessions so feel free to do that for others too

Comment (agriyakhetarpal) [reply noxfile.py:34]: We already have `nox.options.reuse_existing_virtualenvs = True` at the top as a global setting so this argument would not be needed

File: src/pybamm_cookiecutter/parameters/parameter_sets.py (hunk 5 above)
Comment (santacodes) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:117]: Ferran and Agriya wanted to follow the latest PyBaMM release for this project hence I removed that check. You can refer to that discussion [here](https://github.com/pybamm-team/pybamm-cookiecutter/pull/15#discussion_r1649772356).

File: src/pybamm_cookiecutter/parameters/parameter_sets.py (hunk 8 above)
Comment (agriyakhetarpal) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:74]: I don't have enough access to configure it, so I assume you wouldn't either. @brosaplanella, could you help us configure CodeCov? If it helps, here's the link to it: https://github.com/pybamm-team/pybamm-cookiecutter/settings/installations

File: src/pybamm_cookiecutter/parameters/parameter_sets.py (hunk 5 above)
Comment (arjxn-py) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:115]: Didn't saw the other discussion, hence marking this one as resolved.

File: LICENSES-bundled.txt [hunk 12]
```diff
@@ -0,0 +1,7 @@
+This project and source distributions bundle several libraries that are
+compatibly licensed.
+
+
+Name: PyBaMM
```
Comment (brosaplanella) [inline LICENSES-bundled.txt:5]: Given that PyBaMM will install some dependencies, do we need to list them here too?

**Review by brosaplanella — APPROVED** [review]
Just a small comment, otherwise happy to merge if Agriya and Arjun are happy with it too.

File: LICENSES-bundled.txt (hunk 12 above)
Comment (agriyakhetarpal) [reply LICENSES-bundled.txt:5]: I don't think so, because we don't use the source code from those dependencies – it is just PyBaMM.

Comment (agriyakhetarpal) [reply LICENSES-bundled.txt:5]: I don't think so, because we don't use the source code from those dependencies – it is just PyBaMM.

File: LICENSE [hunk 13]
```diff
@@ -1,6 +1,6 @@
 BSD 3-Clause License
 
-Copyright (c) 2023, PyBaMM Team
+Copyright (c) 2023-2024, PyBaMM Team
```
Comment (agriyakhetarpal) [inline LICENSE:3]:
```suggestion
Copyright (c) 2024, PyBaMM Team
```

File: noxfile.py [hunk 14]
```diff
@@ -26,3 +30,11 @@ def run_template_generation(session):
...
+def set_dev(session):
+    """Install pybamm-cookiecutter in editable mode"""
+    session.install("virtualenv")
+    session.run("virtualenv", os.fsdecode(VENV_DIR), silent=True)
+    python = os.fsdecode(VENV_DIR.joinpath("bin/python"))
+    session.run(python, "-m", "uv", "pip", "install", "-e", ".[dev]")
```
Comment (agriyakhetarpal) [inline noxfile.py:40]: I tried this locally and it isn't working – `uv` doesn't seem to properly install dependencies into a `virtualenv`-created environment. I think it needs itself to create the virtual environment, i.e., using `uv venv`. I don't wish to hold the PR, so maybe we can use the original, non-`uv` solution and fix this later?

File: src/pybamm_cookiecutter/parameters/__init__.py [hunk 15]
```diff
@@ -0,0 +1,3 @@
+from __future__ import annotations
+
+__all__ = ["parameter_sets", ]
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/__init__.py:3]:
```suggestion
__all__ = ["parameter_sets",]
```

(nit)

File: src/pybamm_cookiecutter/parameters/input/Chen2020.py [hunk 16]
```diff
@@ -0,0 +1,369 @@
+"""
+This code is adopted from the PyBaMM project under the BSD 3-Clause
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/input/Chen2020.py:2]:
```suggestion
This code is adopted from the PyBaMM project under the BSD-3-Clause
```

The correct SPDX short identifier is "BSD-3-Clause", so either this should be "BSD-3-Clause" as per my suggestion or the entire name, 'BSD 3-Clause "New" or "Revised" License'.

See https://spdx.org/licenses/BSD-3-Clause.html

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
+"""
+This code is adopted from the PyBaMM project under the BSD 3-Clause
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:2]: Same here

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        Parameters for the Enertech cell (Ai2020), from the papers :footcite:t:`Ai2019`,
+        :footcite:t:`rieger2016new` and references therein.
+        ...
+
+    See also: :ref:`adding-parameter-sets`
+
+    """
+
+    def __init__(self):
+        '''Dict of entry points for parameter sets, lazily load entry points as'''
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:66]:
```suggestion
        """Dict of entry points for parameter sets, lazily load entry points as"""
```

I would suggest using double quotes instead of single-quotes to keep consistency with the docstring above this function.

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        self.__all_parameter_sets = dict()
+        for entry_point in self.get_entries("cookie_parameter_sets"):
+            self.__all_parameter_sets[entry_point.name] = entry_point
+
+    @staticmethod
+    def get_entries(group_name):
+        '''Wrapper for the importlib version logic'''
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:73]:
```suggestion
        """Wrapper for the importlib version logic"""
```

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+        >>> list(pybamm_cookiecutter.parameter_sets)
+        ['Chen2020', ...]
+
+    Get the docstring for a parameter set:
+
+
+        >>> print(pybamm_cookiecutter.parameter_sets.get_docstring("Ai2020"))
+        <BLANKLINE>
+        Parameters for the Enertech cell (Ai2020), from the papers :footcite:t:`Ai2019`,
+        :footcite:t:`rieger2016new` and references therein.
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:58]: We don't have doctests or a citations file set up. I think we should set them up after the structure is ready. Let's keep these in mind

File: src/pybamm_cookiecutter/parameters/parameter_sets.py [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+            cls.instance = super().__new__(cls)
+        return cls.instance
+
+    def __getitem__(self, key) -> dict:
+        return self._load_entry_point(key)()
+
+    def _load_entry_point(self, key) -> Callable:
+        """Check that ``key`` is a registered ``cookie_parameter_sets``,
+        and return the entry point for the parameter set, loading it needed.
+        """
```
Comment (agriyakhetarpal) [inline src/pybamm_cookiecutter/parameters/parameter_sets.py:91]:
```suggestion
        """Check that ``key`` is a registered ``cookie_parameter_sets``,
        and return the entry point for the parameter set, loading it needed."""
```

(nit) move the ending quote to the same line for consistency

File: tests/test_entry_points.py [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+import importlib.util
+import sys
+from pathlib import Path
+
+def test_entry_points():
+    """Test if the entry points are loaded correctly."""
+
+    entry_points = list(pybamm_cookiecutter.parameter_sets)
+    parameter_sets = Path("src/pybamm_cookiecutter/parameters/input/").glob("*.py")
+    # Making a list Parameter sets in the parameters/input directory
```
Comment (agriyakhetarpal) [inline tests/test_entry_points.py:12]:
```suggestion
    # Making a list of parameter sets in the parameters/input directory
```

File: tests/test_entry_points.py [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+    """Testing if the values get loaded via parameter entry points and are equal when loaded through entry points"""
+    # Loading parameter_sets through entry points
+    parameters = pybamm_cookiecutter.parameter_sets['Chen2020']
+    # Loading parameter sets through the source file by dynamically loading Chen2020.py as a module
+    spec = importlib.util.spec_from_file_location("Chen2020mod", "src/pybamm_cookiecutter/parameters/input/Chen2020.py")
+    chen_module = importlib.util.module_from_spec(spec)
+    sys.modules["Chen2020mod"] = chen_module
+    spec.loader.exec_module(chen_module)
+    parameters_from_file = chen_module.get_parameter_values()
+    assert parameters.keys() == parameters_from_file.keys(), "The keys in the module and local input file are not the same"
```
Comment (agriyakhetarpal) [inline tests/test_entry_points.py:27]: Could we improve this error message? We can add something like, "Got {}, expected {}". Not sure if it is needed with `pytest`, though.

**Review by agriyakhetarpal — COMMENTED** [review]
Looks good to me too, a few final comments! Thanks, @santacodes!

File: noxfile.py (hunk 14 above)
Comment (santacodes) [reply noxfile.py:40]: Strange, it works for me when I use nox outside the `virtualenv` and it detects the `venv` folder in the working directory and installs the project inside it, perhaps we should look at this later when we are done with model entry points.

**Review by agriyakhetarpal — APPROVED** [review]

File: src/pybamm_cookiecutter/parameters/parameter_sets.py (hunk 8 above)
Comment (arjxn-py) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:74]: Resolving this as being tracked in #19

File: src/pybamm_cookiecutter/parameters/parameter_sets.py (hunk 20 above)
Comment (arjxn-py) [reply src/pybamm_cookiecutter/parameters/parameter_sets.py:58]: This should be covered in #20

File: noxfile.py (hunk 14 above)
Comment (arjxn-py) [reply noxfile.py]: I don't want to hold this PR either, but would it be fine to test this session on CI as well?

Comment (arjxn-py) [reply noxfile.py]: Resolving this, at the same time RFD on #21

**Review by arjxn-py — APPROVED** [review]
Thanks @santacodes, merging this as you must have also started with model entry points so didn't want to hold this.
I hope it didn't raise any bottlenecks.
